### PR TITLE
Context Format for Prompt for Completion & friends

### PR DESCRIPTION
**Pull request by valaises**
Change list:
* for bigcode models: use context from ast in correct format; counting tokens for ast context correctly
* updated format of context field for debug-fim-page to include was_looking_for

### CODE REVIEW

File: src/scratchpads/completion_single_file_fim.rs [hunk 1]
```diff
@@ -71,19 +72,44 @@ impl SingleFileFIM {
...
-        patch: &serde_json::Value,
+        patch: &Value,
     ) -> Result<(), String> {
         // That will work for some models (starcoder) without patching
         self.fim_prefix = patch.get("fim_prefix").and_then(|x| x.as_str()).unwrap_or("<fim_prefix>").to_string();
         self.fim_suffix = patch.get("fim_suffix").and_then(|x| x.as_str()).unwrap_or("<fim_suffix>").to_string();
         self.fim_middle = patch.get("fim_middle").and_then(|x| x.as_str()).unwrap_or("<fim_middle>").to_string();
         self.t.eot = patch.get("eot").and_then(|x| x.as_str()).unwrap_or("<|endoftext|>").to_string();
         self.t.eos = patch.get("eos").and_then(|x| x.as_str()).unwrap_or("").to_string();
+        info!("patch: {}", patch);
```
Comment (valaises) [inline src/scratchpads/completion_single_file_fim.rs:110]: shit, gotta remove this one 😁

File: src/scratchpad_abstract.rs [hunk 2]
```diff
@@ -43,11 +43,13 @@ pub struct HasTokenizerAndEot {
     pub tokenizer: Arc<RwLock<Tokenizer>>,
     pub eot: String,
     pub eos: String,
+    pub supports_context: bool, 
+    pub n_ctx: i32,
 }
```
Comment (olegklimov) [inline src/scratchpad_abstract.rs:48]: `supports_context` -> `context_style`, a string

Comment (olegklimov) [reply src/scratchpad_abstract.rs:48]: n_ctx why do we need one over there?

File: src/scratchpads/completion_single_file_fim.rs [hunk 3]
```diff
@@ -212,20 +207,55 @@ impl ScratchpadAbstract for SingleFileFIM {
...
                 self.fim_prefix,
-                before.into_iter().rev().collect::<Vec<_>>().join(""),
+                before,
                 cursor_line1,
                 self.fim_middle,
             );
         } else {
             return Err(format!("order \"{}\" not recognized", self.order));
         }
+        let prompt_tokens_before_context = self.t.count_tokens(prompt.as_str())?;
```
Comment (olegklimov) [inline src/scratchpads/completion_single_file_fim.rs:223]: Do we really need that? Encoding text is a long operation. We don't have a counter above?

File: src/scratchpads/chat_utils_rag.rs [hunk 4]
```diff
@@ -36,6 +39,24 @@ pub struct FileLine {
...
+        json!({
+            "file_name": x.file_name.clone(),
+            "file_content": x.file_content.clone(),
+            "line1": x.line1.clone(),
+            "line2": x.line2.clone(),
+        })
+    }).collect();
+    
+    let mut res = json!({
+        "was_looking_for": was_looking_for.clone(),
```
Comment (olegklimov) [inline src/scratchpads/chat_utils_rag.rs:53]: "was_looking_for":  [
  {"symbol": "xxx", "something_else": 1337},
  {"symbol": "xxx", "something_else": 1337},
  {"symbol": "xxx", "something_else": 1337}
  ]

File: src/ast/ast_module.rs [hunk 5]
```diff
@@ -182,12 +183,16 @@ impl AstModule {
             top_n_near_cursor,
             top_n_usage_for_each_decl
         ).await;
+        for r in cursor_usages.iter() {
+            let last_30_chars = last_n_chars(&r.name, 30);
+            info!("found {last_30_chars}");
```
Comment (JegernOUTT) [inline src/ast/ast_module.rs:188]: those are not found, but parsed near the cursor

File: src/ast/ast_module.rs [hunk 6]
```diff
@@ -212,38 +217,39 @@ impl AstModule {
...
-                        symbol_declaration: x.clone(),
-                        content: x.get_content_blocked().unwrap_or_default(),
-                        sim_to_query: -1.0,
-                    })
-                    .collect::<Vec<SymbolsSearchResultStruct>>(),
-                matched_by_name_symbols: matched_by_name_symbols
-            }
-        )
+        
+        let was_looking_for = cursor_usages.iter().chain(declarations.iter()).chain(usages.iter()).map(|x|last_n_chars(&x.name, 30)).collect::<Vec<_>>();
```
Comment (JegernOUTT) [inline src/ast/ast_module.rs:221]: Do we really need that here?
We can take symbol names afterward

**Review by JegernOUTT — CHANGES_REQUESTED** [review]

File: src/scratchpads/completion_single_file_fim.rs [hunk 7]
```diff
@@ -212,20 +213,57 @@ impl ScratchpadAbstract for SingleFileFIM {
...
-                before.into_iter().rev().collect::<Vec<_>>().join(""),
+                before,
                 cursor_line1,
                 self.fim_middle,
             );
         } else {
             return Err(format!("order \"{}\" not recognized", self.order));
         }
+        let rag_tokens_n = if self.post.rag_tokens_n > 0 {
+            self.post.rag_tokens_n.min(2048).max(1024)
```
Comment (JegernOUTT) [inline src/scratchpads/completion_single_file_fim.rs:230]: Can we get rid of all hardcoded context sizes?

File: src/ast/ast_module.rs [hunk 8]
```diff
@@ -212,38 +214,43 @@ impl AstModule {
...
-                        symbol_declaration: x.clone(),
-                        content: x.get_content_blocked().unwrap_or_default(),
-                        sim_to_query: -1.0,
-                    })
-                    .collect::<Vec<SymbolsSearchResultStruct>>(),
-                matched_by_name_symbols: matched_by_name_symbols
-            }
-        )
+        
+        let mut was_looking_for = HashMap::new();
```
Comment (JegernOUTT) [inline src/ast/ast_module.rs:218]: move `was_looking_for` outside this function

File: src/scratchpads/completion_single_file_fim.rs [hunk 9]
```diff
@@ -71,19 +75,47 @@ impl SingleFileFIM {
     }
 }
 
+fn add_context_to_prompt(context_format: &String, prompt: &String, postprocessed_messages: &Vec<ContextFile>) -> String {
+    let mut context_files = vec![];
+    if context_format == "starcoder" {
```
Comment (JegernOUTT) [inline src/scratchpads/completion_single_file_fim.rs:80]: Add a default `context_format`
If you cannot figure out which format is the best - just wrap all the code with the comments (see `wrap_comments`) and put it above the prefix part of the code

**Review by JegernOUTT — CHANGES_REQUESTED** [review]

**Pull request merged**